### PR TITLE
PS-1116 : LP #1719506: Audit plugin reports "command_class=error" for…

### PR DESCRIPTION
**Pull request by nitendra-bhosle**
… server-side prepared statements

https://jira.percona.com/browse/PS-1116

Analysis
--------
server-side prepared statements sets
thd->lex->sql_command to SQLCOM_END , which results in
COMMAND_CLASS to set as 'error'

Fix
---
Inside function mysql_audit_notify we have
set COMMAND_CLASS to correct type based on msg type
This change is specifically made for audit logs only

### CODE REVIEW

File: sql/sql_audit.cc
Comment (percona-ysorokin) [inline sql/sql_audit.cc]: There is `command_name` array declared in `sql_parse.h`. You can iterate over its elements and find the index which you then will be able to cast to `COM_XXX` constant. After that you can `switch` over this constant and fill `event.general_sql_command` for known ones
```
  COM_STMT_PREPARE,
  COM_STMT_EXECUTE,
  COM_STMT_SEND_LONG_DATA,
  COM_STMT_CLOSE,
  COM_STMT_RESET,
  COM_STMT_FETCH,
  ...
```
with exactly the same `event.general_sql_command= sql_statement_names[<found_index>];`
This will help to remove `magic constants` from the code

File: sql/sql_audit.cc
Comment (percona-ysorokin) [inline sql/sql_audit.cc]: BTW, here, you could do `event.general_sql_command = C_STRING_WITH_LEN("...")`

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]
Hi @nitendra-bhosle . The fix itself looks good. I have a few questions / requests:

- Where did you get the list of missing statements (stmt_send_long_data, stmt_fetch, ...). The reason for asking is, perhaps there is more statements
- Our documentation mention that the list of commands come from `SELECT name FROM performance_schema.setup_instruments WHERE name LIKE "statement/sql/%" ORDER BY name;`  and the ones you added are from `select * from setup_instruments where name like '%statement/com%';` - https://www.percona.com/doc/percona-server/8.0/management/audit_log_plugin.html#filtering-by-sql-command-type - Documentation needs to be updated to reflect those changes.
- We need test cases showing those commands will be logged with the desired command class
- Test cases must ensure we can filter in/out the new classes

File: sql/sql_audit.cc
Comment (nitendra-bhosle) [reply sql/sql_audit.cc]: I had tried with C_STRING_WITH_LEN("...") ,But I was getting error " error: extended initializer lists only available with ‘-std=c++11’ or ‘-std=gnu++11’ [-Werror]" 
This is warning and warnings are treated as error , so , it's not building with  this change.

May be because of this sql_statement_names is built using 'str' and 'len' members explicitly : 
./sql/mysqld.cc:      sql_statement_names[com_index].str= var->name;
./sql/mysqld.cc:      sql_statement_names[com_index].length= strlen(var->name);
 and 

event.general_command.str= msg;
event.general_command.length= msg_len;

Comment (satya-bodapati) [reply sql/sql_audit.cc]: Please try like this:  = { C_STRING_WITH_LEN("Host") }. (With curly brackets)

Comment (nitendra-bhosle) [reply sql/sql_audit.cc]: I was trying with curly brackets only.
event.general_sql_command = { C_STRING_WITH_LEN("Host") };

is not compiling .

Comment (nitendra-bhosle) [reply sql/sql_audit.cc]: I think inside makefile , CXX flags must be set as : -std=c++03

Comment (nitendra-bhosle) [reply sql/sql_audit.cc]: I have verified this with test program 
g++ test.cpp -o test -std=c++03
test.cpp: In function ‘int main()’:
test.cpp:19:33: warning: extended initializer lists only available with ‘-std=c++11’ or ‘-std=gnu++11’
   19 |  t.test = {"test",strlen("test")};

g++ test.cpp -o test -std=c++11 :  No warning

Comment (satya-bodapati) [reply sql/sql_audit.cc]: How does it work in other places. It is used a lot in 5.7 codebase. Something else is the reason? :thinking:

Comment (nitendra-bhosle) [reply sql/sql_audit.cc]: Inside file : sql_audit.cc , I am not able to use any of c++ 11 features.I am not sure why

Comment (satya-bodapati) [reply sql/sql_audit.cc]: AFAIR, c+11 is from 8.0 code base only. C_STRING_WITH_LEN has nothing do with c+11 feature? :thinking:

Comment (satya-bodapati) [reply sql/sql_audit.cc]: also compare compilation flags of sql_audit.cc with other source files? (use VERBOSE=1 make)

Comment (nitendra-bhosle) [reply sql/sql_audit.cc]: > AFAIR, c+11 is from 8.0 code base only. C_STRING_WITH_LEN has nothing do with c+11 feature?

Yeah right, I was trying to use auto and lamda's

Comment (nitendra-bhosle) [reply sql/sql_audit.cc]: > also compare compilation flags of sql_audit.cc with other source files? (use VERBOSE=1 make)

With VERBOSE=1 , I can see sql_audi.cc is compiling with -std=gnu++03 option

Comment (satya-bodapati) [reply sql/sql_audit.cc]: ok, did you avoid lambdas, autos and C+11 features in 5.7? do you see still same error?

Comment (nitendra-bhosle) [reply sql/sql_audit.cc]: I have figured out issue , with -std=c++03
You can initialise a struct with extended initializer lists , but you can't assign .
That means event.general_sql_command = { C_STRING_WITH_LEN(msg)} ; doesn't work 
But, MYSQL_LEX_CSTRING command_class = { C_STRING_WITH_LEN(msg)} ; works.

File: plugin/audit_log/tests/mtr/prepare_stmt_audit_log.py
Comment (satya-bodapati) [inline plugin/audit_log/tests/mtr/prepare_stmt_audit_log.py]: I think this should be uncommented?

File: plugin/audit_log/tests/mtr/prepare_stmt_audit_log.py
Comment (satya-bodapati) [inline plugin/audit_log/tests/mtr/prepare_stmt_audit_log.py]: remove this

File: plugin/audit_log/tests/mtr/prepare_stmt_audit_log.py
Comment (satya-bodapati) [inline plugin/audit_log/tests/mtr/prepare_stmt_audit_log.py]: cant we use 
PREPARE stmt FROM SHOW GLOBAL VARIABLES WHERE Variable_name = ?;
SET @var2='blah';
EXECUTE insert_stmt USING @var2;

Do we need this python script ?

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
  if (thd->lex->sql_command == SQLCOM_END && msg_len > 0)
```

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
    for (size_t i=0; i < get_command_name_len(); ++i)
```

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
      if (!strcmp(command_name[i].str, msg))
```

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
    if (found_index == -1)
```

File: sql/sql_parse.cc
Comment (satya-bodapati) [inline sql/sql_parse.cc]:
```suggestion
  return sizeof(command_name) / sizeof(command_name[0]);
```

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: plugin/audit_log/tests/mtr/prepare_stmt_audit_log.py
Comment (nitendra-bhosle) [reply plugin/audit_log/tests/mtr/prepare_stmt_audit_log.py]: These statements will generate Audit logs with name "Query" not name "Prepare".
But MySQLdb python driver used for python2 in mtr tests also generates "Query" logs .
So , I will remove python script and use these statements.

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
    int found_index= -1;
```

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
        found_index= i;
```

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
    for (size_t i= 0; i < get_command_name_len(); ++i)
```

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
      event.general_sql_command= sql_statement_names[thd->lex->sql_command];
```

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
        event.general_sql_command= sql_statement_names[SQLCOM_PREPARE];
```

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
        event.general_sql_command= sql_statement_names[SQLCOM_EXECUTE];
```

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
        event.general_sql_command= sql_statement_names[SQLCOM_RESET];
```

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
        MYSQL_LEX_CSTRING command_class= { C_STRING_WITH_LEN(msg)} ;
```

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
        event.general_sql_command= command_class;
```

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
5.7 style is a bit different. Rest all looks good.

File: sql/sql_audit.cc
Comment (satya-bodapati) [inline sql/sql_audit.cc]:
```suggestion
        MYSQL_LEX_CSTRING command_class= { C_STRING_WITH_LEN(msg) } ;
```

**Review by satya-bodapati — APPROVED** [review]
LGTM! one format change requested.

**Review by satya-bodapati — APPROVED** [review]
typo in commit msg: "chnage"

Please wait for @altmannmarcelo's approval before merging

File: plugin/audit_log/tests/mtr/audit_log_charset.result
Comment (altmannmarcelo) [inline plugin/audit_log/tests/mtr/audit_log_charset.result]: Why does this changed from `error` to `Query` ?

From https://github.com/percona/percona-server/blob/5.7/plugin/audit_log/tests/mtr/audit_log_charset.test#L64 we are expecting a Parsing error, and those should be reported as error on audit log.

File: plugin/audit_log/tests/mtr/audit_log_filter_commands.result
Comment (altmannmarcelo) [inline plugin/audit_log/tests/mtr/audit_log_filter_commands.result]: seems like we use lower case for all command classes, and here we are now adding the command classes with the initial upper case.

File: plugin/audit_log/tests/mtr/audit_log_filter_prepare_stmt.test
Comment (altmannmarcelo) [inline plugin/audit_log/tests/mtr/audit_log_filter_prepare_stmt.test:17]: I see you covered the case from https://jira.percona.com/browse/PS-1116. We also have https://jira.percona.com/browse/PS-3530 linked to to it.

If a user runs
1. SELECT field FROM table;
2. PREPARE xxx FROM 'SELECT field FROM table'
3. EXECUTE xxx;

Does item 3 will show up as `execute_sql` command class or `select`?
If I filter by excluding SELECTs, will item 3 show up in audit logs?

File: sql/sql_audit.cc
Comment (altmannmarcelo) [inline sql/sql_audit.cc:461]: just to be on the safe side, any performance issue could arise here by doing a string comparison inside a loop if the workload is mostly done via prepared statements?

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]
@nitendra-bhosle thanks for working on the test cases. A few comments that are still unclear.

File: plugin/audit_log/tests/mtr/audit_log_filter_commands.result
Comment (nitendra-bhosle) [reply plugin/audit_log/tests/mtr/audit_log_filter_commands.result]: These command classes are from query : select * from performance_schema.setup_instruments where name like '%statement/com%';
Which are having initial letter uppercase

File: plugin/audit_log/tests/mtr/audit_log_filter_prepare_stmt.test
Comment (nitendra-bhosle) [reply plugin/audit_log/tests/mtr/audit_log_filter_prepare_stmt.test:17]: Item 3 will show up as 'execute_sql' in audit log ,if you exclude by SELECTS's

File: sql/sql_audit.cc
Comment (nitendra-bhosle) [reply sql/sql_audit.cc:461]: Command classes generated based on 'msg' variable , hence strcmp is used , If we want to avoid this, we need to make changes to sql_parse.cc file

File: plugin/audit_log/tests/mtr/audit_log_filter_prepare_stmt.test
Comment (altmannmarcelo) [reply plugin/audit_log/tests/mtr/audit_log_filter_prepare_stmt.test:17]: This would be a FR, not the context of this bug. Marking as solved.

File: plugin/audit_log/tests/mtr/audit_log_filter_commands.result
Comment (altmannmarcelo) [reply plugin/audit_log/tests/mtr/audit_log_filter_commands.result]: Make it lower case to be consistent.

File: sql/sql_audit.cc
Comment (altmannmarcelo) [reply sql/sql_audit.cc:461]: I'm just wondering that this could be a performance bummer for a workload that is 100% PS (some ORM do this. PHP also heavily uses PS for anti SQL injection).
Is it possible for you to run a quick test with and without your bug, having a workload that is PS (sysbench has it disable by default, but you can enable it https://github.com/akopytov/sysbench/issues/310 )

Comment (nitendra-bhosle) [reply sql/sql_audit.cc:461]: I have run sysbench with PS , I didn't notice any performance degradation. 
I have used these commands : 
sysbench --table-size=1000000 --db-ps-mode='auto' --mysql-db=test --mysql-host=127.0.01.1 --mysql-port=3306 --mysql-user=root --mysql-password=<password> /usr/share/sysbench/oltp_read_write.lua prepare
sysbench --table-size=1000000 --db-ps-mode='auto' --mysql-db=test --mysql-host=127.0.01.1 --mysql-port=3306 --mysql-user=root --mysql-password=<password> /usr/share/sysbench/oltp_read_write.lua run

These are results : 
1. With My patch included : 

Running the test with following options:
Number of threads: 1
Initializing random number generator from current time


Initializing worker threads...

Threads started!

SQL statistics:
    queries performed:
        read:                            2940
        write:                           840
        other:                           420
        total:                           4200
    transactions:                        210    (20.90 per sec.)
    queries:                             4200   (418.01 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          10.0463s
    total number of events:              210

Latency (ms):
         min:                                   37.48
         avg:                                   47.83
         max:                                  106.89
         95th percentile:                       65.65
         sum:                                10044.63

Threads fairness:
    events (avg/stddev):           210.0000/0.00
    execution time (avg/stddev):   10.0446/0.0

2. Without my patch : 
Running the test with following options:
Number of threads: 1
Initializing random number generator from current time


Initializing worker threads...

Threads started!

SQL statistics:
    queries performed:
        read:                            2926
        write:                           836
        other:                           418
        total:                           4180
    transactions:                        209    (20.81 per sec.)
    queries:                             4180   (416.27 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          10.0384s
    total number of events:              209

Latency (ms):
         min:                                   37.20
         avg:                                   48.02
         max:                                   87.09
         95th percentile:                       61.08
         sum:                                10036.33

Threads fairness:
    events (avg/stddev):           209.0000/0.00
    execution time (avg/stddev):   10.0363/0.00

Comment (altmannmarcelo) [reply sql/sql_audit.cc:461]: thanks for testing it

File: plugin/audit_log/tests/mtr/audit_log_echo.inc
Comment (altmannmarcelo) [inline plugin/audit_log/tests/mtr/audit_log_echo.inc]: are Ping and Query suppose to have the initial as upper case here?

**Review by altmannmarcelo — APPROVED** [review]
@nitendra-bhosle thanks for working on those. I have just the observation of the regexp. If required fix it and it's good to go from my side (no need to send for re-review).

File: plugin/audit_log/tests/mtr/audit_log_echo.inc
Comment (nitendra-bhosle) [reply plugin/audit_log/tests/mtr/audit_log_echo.inc]: Ping and Query will not be required here , [a-z_]  will handle that , I will remove these two from here , will keep "init db" so that I don't have to keep random space

**Review by satya-bodapati — APPROVED** [review]

**Pull request merged**
